coast/CLIMATOLOGY.py

```diff
@@ -76,10 +84,122 @@ def make_climatology(
+        # A datetime type is required for slicing data using xarray's sel() method.
+        if not np.issubdtype(time_dim_da.dtype, np.datetime64) and np.issubdtype(time_var_da.dtype, np.datetime64):
+            warn("Time dimension is not np.datatime64 but time variable is. Swapping time dimension for data variable.")
+            new_ds = ds.swap_dims({f"{time_dim}": f"{time_var}"})
+            time_dim = time_var
+        elif np.issubdtype(time_dim_da.dtype, np.datetime64):
+            # Continue without warning.
+            new_ds = ds
+        else:
+            # Slicing will most likely fail for non np.datetime64 datatypes.
+            warn("Neither time dimension or time variable data are np.datetime64. Time slicing may fail.")
+            new_ds = ds
+
+        # Get years of data.
+        data_years = list(new_ds[f"{time_dim}.year"].to_numpy())
+        # Append first year - 1, to account for possible WINTER month data of year prior to data beginning.
+        data_years.insert(0, data_years[0] - 1)
+
+        # Generate date ranges from years and given month periods.
+        date_ranges = CLIMATOLOGY._get_date_ranges(data_years, month_periods)
+
+        # Extract data from dataset between these date ranges and index each range with a common multi-index.
+        datasets = []
+        year_index = []
+        month_index = []
+        for date_range in date_ranges:
+            sel_args = {f"{time_dim}": slice(date_range[0], date_range[1])}
+            filtered = new_ds.sel(**sel_args)
+            datasets.append(filtered)
+            year_index = year_index + ([date_range[0].year] * filtered.sizes[time_dim])
+            month_label = f"{calendar.month_abbr[date_range[0].month]}-{calendar.month_abbr[date_range[1].month]}"
+            month_index = month_index + ([month_label] * filtered.sizes[time_dim])
+
+        # New dataset built from extracted data between date ranges.
+        filtered = xr.concat(datasets, dim=time_dim)
+        # Data from same date range use common year-period multi-index so they can be grouped together.
+        period_idx = pd.MultiIndex.from_arrays([year_index, month_index], names=("year", "period"))
+        filtered.coords["year_period"] = (f"{time_dim}", period_idx)
+
+        # For each data variable, group on year-period multi-index and find the mean.
+        # New dataset containing means across date ranges is returned.
+        ds_mean = xr.Dataset()
+        for var_name, da in filtered.data_vars.items():
+            try:
+                # Apply .mean() to grouped data.
+                # skipna flag used to ignore NaN values.
+                da_mean = da.groupby("year_period").mean(dim=time_dim, skipna=True)
+                ds_mean[f"{var_name}"] = da_mean
+            except ArithmeticError as e:
+                warn(f"Skipped mean calculation for {var_name} due to error: {e}")
+        return ds_mean
+
+
+class Season:
```

Comment:
It might be worth just putting these values in a module rather than using class variables. I think that would be more consistent with Python's naming conventions if we want to mark these as "constants".

Unfortunately as long as we support 3.7 and lower we can't use Final...

Reply:
Moved these into a seaons.py file now so can be imported using from coast import seasons.
Then use as required.

During the redesign I assume the location of this seasons.py file may change, rather than being on the top level of the package.